### PR TITLE
Rspec tests action mailer

### PR DESCRIPTION
**Pull request by magnetron-2000**

### CODE REVIEW

File: app/controllers/tickets_controller.rb [hunk 1]
```diff
@@ -22,8 +22,10 @@ def show
   end
 
   def update
+    old_ticket = {title: @ticket.title.dup, description: @ticket.description.dup}
     if @ticket.update(update_params)
       render json: TicketBlueprint.render(@ticket)
+      UserMailer.with(user: @ticket.worker.user, editor: current_user.worker, old_ticket: old_ticket, new_ticket: @ticket).task_changed.deliver_later
```
Comment (dmitri-wm) [inline app/controllers/tickets_controller.rb:28]: move sending email logic in model to the
`after_update or after_commit` callback (find out the difference)

File: app/controllers/tickets_controller.rb [hunk 2]
```diff
@@ -22,8 +22,10 @@ def show
   end
 
   def update
+    old_ticket = {title: @ticket.title.dup, description: @ticket.description.dup}
```
Comment (dmitri-wm) [inline app/controllers/tickets_controller.rb:25]: use `ActiveModel::Dirty` stuff see https://api.rubyonrails.org/classes/ActiveModel/Dirty.html

File: app/controllers/tickets_controller.rb [hunk 3]
```diff
@@ -37,8 +39,8 @@ def destroy
   end
 
   def state # change ticket state
-    unless current_user.is_admin || current_user.worker.role == "Manager"
-      if current_user.worker.id == @ticket.worker_id
+
+      if current_user.is_admin || current_user.worker.role == "Manager" || current_user.worker.id == @ticket.worker_id
```
Comment (dmitri-wm) [inline app/controllers/tickets_controller.rb:43]: this logic does not belong to controller layer.
There are few ways to refactor it, implement one of:
1) Create on user model method like `can_update_ticket_state?(ticket)` which will contain this logic
2) Create on ticket mode method `state_can_be_change_by?(user)`
3) Use https://github.com/varvet/pundit gem which is very popular abstraction layer for permissions 
My advice is to implement point 3

File: app/controllers/tickets_controller.rb [hunk 4]
```diff
@@ -47,12 +49,13 @@ def state # change ticket state
...
         render json: {message: "you have not access!"}, status: 401
       end
-    end
+
   end
 
   def change_worker # change ticket worker
     if @ticket.update(worker_id: params[:worker_id])
       render json: TicketBlueprint.render(@ticket)
+      UserMailer.with(user: @ticket.worker.user, ticket: @ticket).assigned_new_task.deliver_later
```
Comment (dmitri-wm) [inline app/controllers/tickets_controller.rb:58]: move this logic to model level

File: app/controllers/users/registrations_controller.rb [hunk 5]
```diff
@@ -19,9 +19,10 @@ def register_success
       user: current_user,
       worker: current_user.worker
     }, status: :ok
+    UserMailer.with(user: current_user).welcome_email.deliver_later
```
Comment (dmitri-wm) [inline app/controllers/users/registrations_controller.rb:22]: move this logic to  User model level in `after_create` callback

File: app/mailers/application_mailer.rb [hunk 6]
```diff
@@ -1,4 +1,4 @@
 class ApplicationMailer < ActionMailer::Base
-  default from: 'from@example.com'
+  default from: 'danikfox1616@gmail.com'
```
Comment (dmitri-wm) [inline app/mailers/application_mailer.rb:2]: this is out of current task scope but this is good starting point to meet and understand how to serve different values in different enviroments.  There are also few approaches 
1) Use https://github.com/bkeepers/dotenv
2) https://edgeguides.rubyonrails.org/security.html#custom-credentials
Personally I prefer point 1)

File: spec/requests/admins_spec.rb [hunk 7]
```diff
@@ -2,23 +2,46 @@
...
-      before(:each) do
-        post '/users/sign_in', :params => {
-          "email" => "asdfdf@mail.com",
-          "password"=> "secret",
-          "password_confirmation"=> "secret"
-        }
+    context 'add_to_admins and remove from admins' do
+
+      before do
+        post '/users', :params => { "user": {
```
Comment (dmitri-wm) [inline spec/requests/admins_spec.rb:8]: do we need those requests here?

File: spec/requests/admins_spec.rb [hunk 8]
```diff
@@ -2,23 +2,46 @@
...
+
+        user = User.last
+        user.is_admin = true
+        user.save
       end
 
-      it 'should return 200 http status code' do
-        patch '/add_admin/2'
+      it 'should return 200 http status code if user become admin' do
+        patch '/add_admin/1'
```
Comment (dmitri-wm) [inline spec/requests/admins_spec.rb:35]: do not use hardcoded `id` values
should look like


``` 
     let(:user) { create(:user) }

     it 'should blablabla' do
          patch "/add_admin/#{user.id}"
          ...
      end
```

File: spec/requests/admins_spec.rb [hunk 9]
```diff
@@ -2,23 +2,46 @@
...
         expect(response).to have_http_status(:ok)
+        expect(JSON.parse(response.body)).to eq({"message"=>"dfddfdfddsfd@mail.com is admin"})
       end
 
-      it 'is admin true' do
-        patch '/add_admin/2'
-        expect(User.find_by(id: 2)).is_admin.to eq(true)
+      it 'should return 200 if user lost admin status' do
+        patch '/add_admin/1'
```
Comment (dmitri-wm) [inline spec/requests/admins_spec.rb:41]: same. as above

File: spec/requests/registrations_spec.rb [hunk 10]
```diff
@@ -3,34 +3,35 @@
...
   describe "#create" do
     context 'when user exist' do
-      let(:user) {create :user}
-
+      let(:user) {build(:user)}
       it 'should return 201 http status code' do
-        post '/users', :params => {  "email" => user.email,
-                                     "password"=> user.encrypted_password,
-                                     "password_confirmation"=> user.encrypted_password}
+        post '/users', :params => { "user": {
```
Comment (dmitri-wm) [inline spec/requests/registrations_spec.rb:8]: move sended params into `let(:params)`

File: spec/requests/sessions_spec.rb [hunk 11]
```diff
@@ -3,25 +3,29 @@
...
-               "password"=> user.encrypted_password,
-               "password_confirmation"=> user.encrypted_password
-             }
+      it 'should return 200 http status code if user sign in' do
+        post '/users/sign_in', :params => { "user": {
+                                            "email": "dfdsddsfd@mail.com",
+                                            "password": "secret",
+                                            "password_confirmation": "secret" }}
+        expect(response).to have_http_status(:ok)
+        # expect(JSON.parse(response.body)).to eq({"message"=>"You are signed out."}) #TODO invalid response
```
Comment (dmitri-wm) [inline spec/requests/sessions_spec.rb:12]: cleanup?

File: spec/requests/sessions_spec.rb [hunk 12]
```diff
@@ -3,25 +3,29 @@
 RSpec.describe '/session controllers' do
   describe "#create, delete" do
     context 'post request sign in' do
```
Comment (dmitri-wm) [inline spec/requests/sessions_spec.rb:5]: `context` or `describe` should describe only one action so each `create` and `delete` should have separate context

File: spec/requests/tickets_spec.rb [hunk 13]
```diff
@@ -3,9 +3,34 @@
 RSpec.describe TicketsController do
   describe "tickets controller" do
     HEADERS = { "ACCEPT" => "application/json" }
-
     let(:ticket) {create(:ticket)}
 
+    before do
+      post '/users', :params => { "user": {
+        "email": "dfddfdfddsfd@mail.com",
```
Comment (dmitri-wm) [inline spec/requests/tickets_spec.rb:10]: move params into `let(:params)`

File: spec/requests/workers_spec.rb [hunk 14]
```diff
@@ -3,8 +3,32 @@
 RSpec.describe WorkersController do
   describe "worker controller" do
     HEADERS = { "ACCEPT" => "application/json" }
-    let(:worker) {create(:worker)}
-    let(:ticket) {create(:ticket, worker:worker)}
+
+    before do
+      post '/users', :params => { "user": {
```
Comment (dmitri-wm) [inline spec/requests/workers_spec.rb:8]: do we need these requests ? can we do it on db level?

File: spec/requests/workers_spec.rb [hunk 15]
```diff
@@ -16,49 +40,40 @@ def tickets_count(worker)
...
-
-    context "when ticket exist" do
-      before do
-        worker
-      end
-
+    context "when worker exist" do
       it 'return valid json index' do
         get'/workers'
         body = JSON.parse(response.body)
```
Comment (dmitri-wm) [inline spec/requests/workers_spec.rb:46]: I think you already impemented helper like `json_response` or smth like it?

**Comment by dmitri-wm** [comment]
Also I don't see tests which covers `permissions` logic

File: app/controllers/tickets_controller.rb (hunk 1 above)
Comment (dmitri-wm) [inline app/controllers/tickets_controller.rb:28]: move this logic to model level

File: app/mailers/application_mailer.rb (hunk 6 above)
Comment (magnetron-2000) [reply app/mailers/application_mailer.rb:2]: 'dotenv-rails' is not available in 'asdf: 2.6.3'
second version also don't work  - command not found

File: spec/requests/admins_spec.rb (hunk 8 above)
Comment (magnetron-2000) [reply spec/requests/admins_spec.rb:35]: expected: {"message"=>"dfddfdfddsfd@mail.com is admin"}
     got: {"errors"=>"already admin!"}
I really don't understand why it change id for current user, not for target

File: spec/requests/workers_spec.rb (hunk 15 above)
Comment (magnetron-2000) [reply spec/requests/workers_spec.rb:46]: blueprint?

File: spec/requests/workers_spec.rb (hunk 14 above)
Comment (magnetron-2000) [reply spec/requests/workers_spec.rb:8]: let user don't work

**Pull request closed without merging**